Misc documents based on "JSAPI User Guide"

ptomato opened the pull request:
Another batch based on the hopefully still-salvageable parts of a page from MDN. Let me know if everything here checks out.

ptomato committed:
docs: Add documents based on JSAPI User Guide from MDN

This is based on the original article at
https://developer.mozilla.org/en-US/docs/Mozilla/Projects/SpiderMonkey/JSAPI_User_Guide
converted to Markdown and split into several files.

Many sections were already covered in one of the existing examples or
markdown files; obsolete; irrelevant; or too rambling. These were
removed.

ptomato committed:
cookbook: Add example for JS_InitClass

This is taken from the "JSAPI User Guide" page and fleshed out a bit
compared to that.

ptomato commented on docs/Custom Objects.md:
```diff
@@ -0,0 +1,127 @@
+
+In either case, if you create an object and then want it to persist in
+the run time where it can be used by other scripts, you must root the
+object by calling `JS_AddRoot` or `JS_AddNamedRoot`.
```
Oops, this does not apply anymore.

hotsphink replied:
Yeah, you could probably swap in JS::PersistentRooted<T> for these. Though it makes me nervous to recommend for much of anything; 90% of the uses of PersistentRooted end up creating unwanted leaks. It would almost be better to have a single PersistentRooted<MyGlobalStuff>, and have MyGlobalStuff be a struct that holds everything you need (with a trace() method that keeps it all alive.)

But anyway, for now it'd probably be fine to just talk about PersistentRooted here.

hotsphink replied:
...or even to suggest installing these sorts of things as named properties on the global, and not using PersistentRooted at all, perhaps?

ptomato replied:
In GJS we use `JS_AddExtraGCRootsTracer` to trace things from the context, is that a reasonable approach?

hotsphink replied:
Yeah, that's totally reasonable and is probably more straightforward for many embeddings -- "do this to identify the embedding roots" as opposed to "this structure contains the embedding roots". I think both can make sense, it's dependent on the situation.

hotsphink reviewed (approved):
I skimmed through this stuff and it seemed more or less reasonable.

ptomato committed:
docs: Replace references to JS_CompileScript and JS_EvaluateScript

Instead, use the more modern APIs JS::Compile and JS::Evaluate.

ptomato committed:
docs: Use JS_NewPlainObject to create an object without JSClass

JS_NewObject takes a JSClass pointer, and the other two mentioned APIs
no longer exist. Since the text mentions creating an object without a
JSClass, JS_NewPlainObject is the right way.

ptomato committed:
docs: Remove references to JS_AddRoot

This is outdated. The better way to make an object available to a script
would be to define it on the global.

ptomato commented:
Thanks for the review. I made that tweak and removed a couple more things that I realized were outdated. I'll merge this now.
